BUG: fix read_csv to parse timezone correctly

swyoon opened the pull request:
Use `box=True` for `to_datetime()`, and adjust downstream processing to the change.

- [x] closes #22256
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

swyoon commented:
[pandas.io.parsers._infer_types](https://github.com/pandas-dev/pandas/blob/b5d81cfe43eeccfc3641aa9578097f726da9ce9d/pandas/io/parsers.py#L1607) breaks. Specifically, [pandas._libs.parsers.sanitize_objects](https://github.com/pandas-dev/pandas/blob/b5d81cfe43eeccfc3641aa9578097f726da9ce9d/pandas/_libs/parsers.pyx#L2230) only accepts ndarray as the input. Otherwise, it raises an exception.

jreback commented:
right so just put a
np.asarray right before sanitize (this doesn’t copy it already an ndarray)

and see what happens

swyoon commented:
We if cast data into np.ndarray, it will lost the timezone information. ndarray can't contain timezone information. [link](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.datetime.html#changes-with-numpy-1-11) That's why I had to bypass it.

mroeschke commented:
So this is a case I didn't account for when I recently fixed `to_datetime` parsings of offsets.

```
In [1]: data = ['2018-01-04 09:01:00+09:00', '2018-01-04 09:02:00+09:00']

# This output should probably behave like In[5]
In [3]: pd.to_datetime(data, box=False)
Out[3]:
array(['2018-01-04T00:01:00.000000000', '2018-01-04T00:02:00.000000000'],
      dtype='datetime64[ns]')

In [4]: data = ['2018-01-04 09:01:00+09:00', '2018-01-04 09:02:00+08:00']

In [5]: pd.to_datetime(data, box=False)
Out[5]:
array([datetime.datetime(2018, 1, 4, 9, 1, tzinfo=tzoffset(None, 32400)),
       datetime.datetime(2018, 1, 4, 9, 2, tzinfo=tzoffset(None, 28800))],
      dtype=object)
```

I don't have the patch off the top of my head, but what would happen if instead if the `to_datetime` call returned an array of datetime objects? If it returns the expected result, the fix would probably be in `to_datetime` instead of the parsing code.

swyoon commented:
@mroeschke I don't want to modify `to_datetime`. Since the `box=True` means the return of `to_datetime` should be an index, so I think `to_datetime` should remain as is. I just want to update CSV parsing process.

I think I have come up with a solution.
```
In [1]: data = ['2018-01-04 09:01:00+09:00', '2018-01-04 09:02:00+09:00']
In [2]: dti = pd.to_datetime(data, box=True)
In [3]: np.array(dti.tolist())
Out[3]:
array([Timestamp('2018-01-04 09:01:00+0900', tz='pytz.FixedOffset(540)'),
       Timestamp('2018-01-04 09:02:00+0900', tz='pytz.FixedOffset(540)')],
      dtype=object)

```
The whole problem was caused by `numpy.ndarray` with `datetime64` dtype not being able to contain timezone information. However, `numpy.ndarray` with `object` dtype can contain pandas `Timestamp`s. This makes the PR much simpler.

jreback commented:
@swyoon object arrays of Timestamps are quite non performant
this is why we use DTI in the first place

swyoon commented:
@jreback Right. I was also worried about the performance. Then should fall back to the first version of the PR.
btw, `np.asarray` to a datetime index will result in `datetime64` dtype, resulting in lost of timzone.

swyoon commented:
@jreback all checks green. Could you please take a look? The change in code is minimal.

swyoon commented on pandas/io/parsers.py:
```diff
@@ -3030,14 +3033,19 @@ def converter(*date_cols):
                     errors='ignore',
                     infer_datetime_format=infer_datetime_format
                 )
+                if not isinstance(converted, DatetimeIndex):
```
`box=True` will make a non-datetime array into an `Index`, resulting downstream errors. The downstream expects `ndarray` rather than `Index`

swyoon committed:
BUG: fix read_csv to parse timezone correctly

- use box=True for to_datetime(), and adjust downstream processing to
the change.
- resolve #22256

swyoon commented:
@jreback I have applied `np.asarray` as you requested. It works well :) All checks are green.

gfyoung commented on pandas/tests/io/parser/parse_dates.py:
```diff
@@ -674,3 +674,19 @@ def test_parse_date_float(self, data, expected, parse_dates):
         tm.assert_frame_equal(result, expected)
+
+    def test_parse_timezone(self):
+        import pytz
```
Move import to top of file.

gfyoung commented:
@swyoon : Can you make the changes in #22422?  They should work with your PR.

swyoon committed:
CLN: Remove try-except in parse_dates test

swyoon commented:
@gfyoung Added a commit to fix #22422. It worked fine. 
Travis seems to fail occasionally, and it doesn't seem to be related to my PR. Could you take a look?

gfyoung commented:
Yeah...Travis has been acting up a bit.  Restarted the failing builds, as they didn't look related to this PR.

swyoon commented:
@gfyoung Thank, but it failed again... [failed job](https://travis-ci.org/pandas-dev/pandas/jobs/418065455)  Could you run it one more time, please?

swyoon commented:
@gfyoung Sorry, but travis still doesn't work

jreback reviewed (changes requested):
small comment, lgtm. ping on green.

jreback commented:
I *think* we also might have another open issue about, if you'd do a search and see.

swyoon commented:
@jreback @gfyoung Finally, it's green. This build took so long. I will check out open issues that might be relevant to this PR.

swyoon commented:
@jreback @gfyoung could you please take a look?

gfyoung reviewed (approved):
LGTM!

cc @jreback

jreback commented:
thanks @swyoon

swyoon commented:
@jreback @gfyoung @mroeschke Thank you all
